### DISCUSSION TITLE
Add a folder

### DISCUSSION THREAD

**Discussion by mbrasseau** [community/community#22534] [New to GitHub]
<p>How do we add a folder to a repository?   I have searched and searched and I see an icon for a new file, but nothing anywhere for a new folder.  Nor did I find the question asked anywhere here on the forum, which I find quite weird.</p>

**Comment by SethClydesdale** [comment, marked as answer]
<p>Hello <a class="mention" href="/mbrasseau">@mbrasseau</a>,</p>
<p>To create a new folder in a repository click “create a new file.” Type your new folder’s name in the area where you would write the file name, and at the end of the file name type a “/” to initilize it as a folder. After this you can create a new file in the folder.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/d8beec8d932fb0fa855667a9e411537b3a08190d.gif" data-download-href="/uploads/short-url/uVqbyOse6CndkQK6qj9hOsyUGXX.gif?dl=1" title="tD3c2Td" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/10122549/181071679-d6657c05-4373-4bad-95d0-6b754941bfbd.gif" alt="tD3c2Td" data-base62-sha1="uVqbyOse6CndkQK6qj9hOsyUGXX" width="690" height="386" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/d8beec8d932fb0fa855667a9e411537b3a08190d_2_690x386.gif, https://aws1.discourse-cdn.com/github/original/2X/d/d8beec8d932fb0fa855667a9e411537b3a08190d.gif 1.5x, https://aws1.discourse-cdn.com/github/original/2X/d/d8beec8d932fb0fa855667a9e411537b3a08190d.gif 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/d8beec8d932fb0fa855667a9e411537b3a08190d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">tD3c2Td</span><span class="informations">1020×572 88.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Alternatively, you can clone the repo locally and add a new folder, and then push the changes to github using the <a href="https://help.github.com/en/articles/adding-a-file-to-a-repository-using-the-command-line" rel="nofollow noopener">command line</a> or <a href="https://desktop.github.com/" rel="nofollow noopener">Github Desktop</a>.</p>

**Reply by Omi98** [reply]
Thanks for this answer. Saved my day :)

**Reply by OzgunMunar** [reply]
Since I developed a Asp. Net Web Api with more than 100 files and GitHub doesn't accept uploading more than 100 files, this answer saved my day. Thanks for the help.

**Reply by giamarieworks** [reply]
Thank you so much for this answer.

**Reply by Glorycs29** [reply]
Thank you for answering

**Reply by CyberRaA2c4e5t** [reply]
Nice

**Reply by brotheralameen1** [reply]
Thanks

**Reply by ZaurKurbanismailov-Azur** [reply]
Thanks you!

**Reply by dan-ie** [reply]
thanks

**Reply by Baonguyen1711** [reply]
Omg thank you so much

**Reply by ZYMNZ** [reply]
Ngl, it would be wayyy easier if they can add it as feature or as a command ><

**Reply by HolasoyPeteRich** [reply]
Ey, thanks man. You helped a lot! 👍 ;)

**Reply by kishan-barai** [reply]
Thank you!

**Reply by iljester** [reply]
I agree. Surely you know how fun it is to load scripts with dozens of subfolders. All created by hand one by one... If there is a simpler solution, I'm waiting.

**Reply by ipan-im** [reply]
Thank You

**Reply by Abeautifulsnow** [reply]
Cool!!! I never know it  "type a “/” to initilize it as a folder." ever before. Thxxx!

**Reply by 1alexander3** [reply]
I was struggling to figure this out, thank you!

**Reply by shivaram235vemula** [reply]
thanks you really helped.

**Reply by naveen3830** [reply]
Thankyou for helping out!

**Reply by Nitish252000** [reply]
Thanks

**Reply by satyakam-vats** [reply]
thank you

**Reply by ADITISHUKLA0111** [reply]
This thread helped me. Thanks

**Reply by frenchfrysays3** [reply]
doesn't work

**Reply by chandrakiran-malkapuram** [reply]
Thankyou for Helping

**Reply by michaelbapst** [reply]
Many many Thanks

**Reply by marcaocamargo** [reply]
thank you so much

**Reply by bcfletcher** [reply]
While that is a neat trick, it is a 100% failure from good user interface design perspective. There should be a menu option to create a directory.

**Reply by Richard-Sebastian-C** [reply]
Thanks for the help

**Reply by Mattu104** [reply]
and to make an empty folder just use a .gitignore

**Reply by Arjunr3108** [reply]
super nice work around but a simple add folder from GitHub would have made that simple, Thanks.

**Reply by Gamer-Within001** [reply]
Thank You

**Reply by NYEEH** [reply]
> Hello [@mbrasseau](/mbrasseau),
> 
> To create a new folder in a repository click “create a new file.” Type your new folder’s name in the area where you would write the file name, and at the end of the file name type a “/” to initilize it as a folder. After this you can create a new file in the folder.
> 
> [<img alt="tD3c2Td" width="690" height="386" src="https://user-images.githubusercontent.com/10122549/181071679-d6657c05-4373-4bad-95d0-6b754941bfbd.gif"> tD3c2Td1020×572 88.9 KB ](https://aws1.discourse-cdn.com/github/original/2X/d/d8beec8d932fb0fa855667a9e411537b3a08190d.gif)
> Alternatively, you can clone the repo locally and add a new folder, and then push the changes to github using the [command line](https://help.github.com/en/articles/adding-a-file-to-a-repository-using-the-command-line) or [Github Desktop](https://desktop.github.com/).

Nearly 2025 and still hasn't changed. What an embarrassment.

**Reply by Mursteinen** [reply]
Thats how to make a file AND a folder, not how to make a folder.
In order to make only a folder, you now have to delete that file too.

**Reply by WriteRead024** [reply]
this Create New File with folder name prefix method still works ok but still requires a placeholder file

to delete the placeholder file use the ellipsis (...) button near the top of the file area when looking at the file and file contents on the GitHub website for the file.  The delete file function then is/was at the bottom of the ellipsis menu

**Reply by malik78945** [reply]
Or you can just rename the file and in the rename box, add the name of the folder followed by /. For example, if your file is newFile.md, just rename it newFolder/newFile.md. 
Works fine

**Reply by demironmanx** [reply]
> Thats how to make a file AND a folder, not how to make a folder. In order to make only a folder, you now have to delete that file too.

Github simply doesn't want empty folders. When you delete the last file in a folder that folder also will be deleted!

**Reply by semogacybernetix** [reply]
What? Github doesn't want empty folders? And maybe Github doesn't want files with the name blubb, so every file that has the name blubb will deleted. I want to add a directory tree with thousends of directorys and subdirectorys and files. It makes no sense to create all directories manually and put the files in it.

**Comment by mbrasseau** [comment]
<p>Thank you so very much!</p>

**Reply by khalmam** [reply]
Thank you so very much!

**Comment by mukulmc03** [comment]
<p>Hello, thanks for the solution! What if I want to create a new folder then upload files from the desktop?</p>

**Reply by MoosecianImmerse** [reply]
create a random file. Then go to folder an add desktop files. Thene delete random file

**Reply by Glorycs29** [reply]
You can add files from your local device from the option: add file and then upload files.

**Reply by Komalprasad132** [reply]
but I want to add a folder instead of a file from my local machine, please guide

**Reply by Giannasaurus** [reply]
Up^

**Comment by SethClydesdale** [comment]
<p>Hello <a class="mention" href="/mukulmc03">@mukulmc03</a>, welcome to the GitHub Support Community!</p>
<p>You can also create new folders by simply dragging and dropping a folder with files to your repository on GitHub and then uploading them. If you want the folder you’re uploading to be inside a specific sub-folder, be sure to click on and enter that sub-folder before uploading your files.</p>

**Reply by NarendraShalivahana** [reply]
thanks man drag n drop easy peesy

**Reply by WriteRead024** [reply]
This upload method still works useful for uploading many files in folder(?s) but still requires a placeholder file ok like the new file on the website method did.

**Comment by ghost** [comment]
<p>The suggestion to name a directory as <code>directory/</code> no longer appears to work. The <code>create File</code> menu item insists on a file being added to the directory. The question remains, how to create an empty directory in a GitHub repository.</p>

**Reply by kayysean** [reply]
drag and drop solution still works. adding a dummy readme file when creating the dir also works.

**Reply by Nathaanlennon** [reply]
drag an empty folder doesnet work anymore

**Comment by airtower-luna** [comment]
<p>You can’t, that’s part of how git works. Git does not track directories other than as parent directories of tracked files.</p>
<p>If you want a directory to exist on checkout, you can create an empty file in that directory (common filenames for that purpose would be <code>.keep</code> or <code>.gitkeep</code>). However, in many cases it’ll be better to let the build process for your project create the directory if necessary.</p>

**Reply by WriteRead024** [reply]
My "create an empty folder" use situation was to create a folder into which then upload an image file.  That can be done by the "upload a folder with file(s) content" method, so really don't need empty GitHub repository folders yet exactly.

**Reply by aflacake** [reply]
You must create a folder with files, otherwise. Folder not created

**Comment by ghost** [comment]
<p>Thank you, it helped me to get started with Github in the way I wanted to.</p>

**Comment by lunapari455** [comment]
<p><strong>On GitHub you can do it this way:</strong></p>
<ol>
<li>Go to the folder inside which you want to create another folder.</li>
<li>Click on New file.</li>
<li>On the text field for the file name, first write the folder name you want to create.</li>
<li>Then type / . …</li>
<li>You can add more folders similarly.</li>
<li>Finally, give the new file a name (for example)<br>
<a href="https://stackoverflow.com/questions/12258399/how-do-i-create-a-folder-in-a-github-repository" class="inline-onebox" rel="noopener nofollow ugc">How do I create a folder in a GitHub repository? - Stack Overflow</a><br>
Also You can read full article on here: <a href="https://www.hairnode.com/ornate-hair-clipper-review/" rel="noopener nofollow ugc">website</a>
</li>
</ol>

**Comment by fungos34** [comment]
I'm new to GitHub, so maybe i didn't get it yet. But to me this seems to be a crucial flaw of this platform. How can a coding platform which is so widely used not support such an essential functionality ...

**Reply by PocroLoco** [reply]
I agree that the way they do folders is kind of odd. I'm also new to GitHub but I dont get why you cant just add a create folder button like you would on like windows file explorer or any file system for that matter.

**Reply by Hyperion72R** [reply]
The problem of missing empty folders is related to git's data structures, look at this:
- https://www.quora.com/Why-is-Git-so-awful/answer/Steven-J-Owens
- https://archive.kernel.org/oldwiki/git.wiki.kernel.org/index.php/GitFaq.html#Can_I_add_empty_directories.3F

**Reply by ed2050** [reply]
Then github should put an empty file in it.  Honestly it's not that hard to implement something sensible, instead of pushing the problem onto confused users with no idea what to do.  Make a **** folder button already.

**Comment by MoosecianImmerse** [comment]
yw

**Comment by Aksshay88** [comment]
To add a folder to a GitHub repository, you can follow these general steps:

Navigate to Your Repository:
Open the GitHub repository where you want to add the folder.

Upload Files:

If the folder you want to add doesn't exist locally on your machine, create it using your file explorer or command line.
If the folder already exists on your machine, make sure it contains the files you want to add.
Commit Changes:

Open a terminal or command prompt.

Navigate to the local repository's directory using the cd command.

Use the following commands to add the folder and its contents, and commit the changes:

bash
```
git add foldername/*   # Replace 'foldername' with the actual name of your folder
git commit -m "Add a descriptive commit message"
```
This stages the changes (using git add) and then commits them with a message.

Push Changes to GitHub:

After committing changes locally, you need to push them to the GitHub repository.

bash
```
git push origin branch_name
```
Replace 'branch_name' with the name of the branch you are working on (often, it's the 'main' branch).

Now, your folder and its contents should be added to the GitHub repository.

If you are working with a new repository or need to initialize a new repository on GitHub, you can follow GitHub's instructions when creating a new repository, including a guide on initializing with a README file and adding existing files and folders. You can find these options when creating a new repository on the GitHub website.

**Reply by Argo624** [reply]
Thanks. It worked.

**Reply by labrat08** [reply]
Perfect! Thank you!

**Comment by bcfletcher** [comment]
whomever at github thought that adding a "/" to a filename was an appropriate way to create a folder FLUNKED user interface design 101. That is the most asinine non-intuitive poorly thought out way to create a folder i have ever seen in 40 years of software development. 

it is basically stupid.

thanks @SethClydesdale  for writing up the answer. it is rather unfortunate that you had to do because of github's poor UI design

**Comment by manan307** [comment]
your file_name "/" and your folder created

**Reply by bcfletcher** [reply]
> your file_name "/" and your folder created

I understand you can do that. My point is that is a massive user interface design failure when it is the only way to create the folder. While a convenient feature, in well thought thru design there  would be an explicit menu item to also create folders. GitHub's web interface has a large number of examples of poor and incoherent design and clearly no one experienced  in good UI design was involved.

**Reply by manan307** [reply]
Agree, you can do drag and drop also

**Comment by RK-CBE** [comment]
Thanks

**Comment by Mikolaj100** [comment]
Thanks #1

**Comment by ghost** [comment]
Thank you so much

**Comment by vinkay215** [comment]
To create a new folder in a repository click “create a new file.” Type your new folder’s name in the area where you would write the file name, and at the end of the file name type a “/” to initilize it as a folder. After this you can create a new file in the folder.

**Comment by ghost** [comment]
not working

**Reply by aflacake** [reply]
Try use `/` want to make folder or directory. Go to folder (created) if you want to create some file

**Comment by rubyFeedback** [comment]
Hmm. This one confused me just a moment ago, because I was searching for "create a new directory". So I was not used to "create a new file" being used for this. Also, folder is a strange name - should we not call it directory instead?

Right now this confuses me. But if everyone else is using that terminology here, well ...

Still, UI-wise I would actually add a second entry, "add a new directory". Right now it is a bit confusing that this has to be done via "add a new file" - I was confused why we can only add new files but not a new directory! Of course we can via appending the /, but I still find that confusing when it is meant to create a FILE. I understand we can create a new file, and automatically create a directory, but what if I FIRST want to create the directory, and only lateron add a file? On Linux I often create a dir and only later add content, such as by writing a new file. I never ever FIRST create the file and only upon SAVING that file, expect that the directory is AUTOMATICALLY created. That seems backward logic to me and I think many Linux users here also agree with that - see how many upvotes the original threadstarter got.

In a comment above it is reasoned by some that this is due to git not tracking directories. This may be, but github is more than "merely" git; it also lets people view files and directories content easily. So github already uses an UI and some kind of internal logic that is independent on git, so based on that rationale and observation I would also reason that "add new directory" should be an action too, irrespective over as to whether git itself tracks directories or not.

**Comment by Richard-Sebastian-C** [comment]
They need to make a better way to make folders. Its 2025

**Reply by bcfletcher** [reply]
I 100% agree the current approach is archaic, non-intuitive, and a major example of bad user interface design.  the fact that this entire issue is even here proves their UI design is flawed.

**Reply by aflacake** [reply]
Use CLI is faster than HTTPS. I also difficult to edit files here (use HTTPS)

**Reply by ComotoJC** [reply]
This is preposterous really.  We had a repo get a little tangled up because old employees forgot to push some things blah blah stuff happens.   Not a big deal.  But a new folder was added inside the repo.  Still not a big deal.  GHDT Windows finds and shows the folder.  Stages a commit.  Basically "hey look I found this folder don't you want to commit it and the files inside?"  Seems absolute childs play. 

But NO.  CANT DO IT.   Have to download and install git bash CLI and jump through idiotic hoops ("**git add .**" by the way, which isn't mentioned in almost any of the answers to people having this problem... "git add foldername/*" BLOWS because it will create either a submodule or an embedded repo).  

How GHDT can find the folder and files and yet not offer the simply GUI functionality to "do you want to add this folder?" (or "make submodule" as an option even) is ludicrous.  What a waste of time for an insanely obvious situation and operation.

**Reply by aflacake** [reply]
If you want remote repository

**Comment by aflacake** [comment]
Open your repo and create a new file, give a name to your folder, example you want 3 folders, like:
`folder1/folder2/folder3/`
after deciding the folder give the file:
`folder1/folder2/folder3/file-file.txt`

The file `file-file.txt` locate folder3. When you create file at folder1, go to folder1/folder2 > navigate to folder1

**Reply by Mursteinen** [reply]
You are explaining how to make a file, not how to make a directory.

Its a Chat GPT level answer.

**Reply by aflacake** [reply]
Folder called "directory" and I dont use chat GPT for the answer 🙂

**Comment by MelanieS** [comment]


**Reply by aflacake** [reply]
Thanks!

**Comment by MelanieS** [comment]
This took me an embarrassingly long time to figure out, but I created a guide on how to do it once I figured it out: https://youtu.be/IeGElVLEbao

I hope it helps someone :)

**Comment by AprilEchoes** [comment]
I don't even see a new file

**Comment by Qasimraza110** [comment]
Open your repository on GitHub.

Click “Add file” → “Create new file”.

In the file name field, type your new folder’s name followed by a / (e.g., myFolder/).

After the slash, type the name of the first file you want to create inside that folder.

Add your file content, then commit the changes.

**Comment by Trasscoon** [comment]
I can't believe github can't even bother to create a such a simple feature.
